### ISSUE TITLE
Borrow checking for static methods became more strict

### ISSUE THREAD

**Issue by Andrepuel** [rust-lang/rust#100725]
### Code

I tried this code:

```rust
pub struct Bigger<'a> {
    _marker: &'a (),
}
impl<'a> Bigger<'a> {
    pub fn get_addr(byte_list: &'a mut Vec<u8>) -> &mut u8 {
        byte_list.iter_mut().find_map(|item| {
            Self::other(item);
            Some(())
        });

        byte_list.push(0);
        byte_list.last_mut().unwrap()
    }

    pub fn other<'b: 'a>(_value: &'b mut u8) {
        todo!()
    }
}
```

I expected to see this happen: I don't know what should be the behavior. Looks like it all boils down to what is going to be `'a` within `Bigger::<'a>::other`. In stable version of Rust, the code was accepted.

Instead, this happened: Using nightly version of the compiler a borrow checker error happened.

### Version it worked on

It most recently worked on: cargo 1.63.0 (fd9c4297c 2022-07-01) / rustc 1.63.0 (4b91a6ea7 2022-08-08)

### Version with regression

`rustc --version --verbose`:
```
rustc 1.65.0-nightly (9c20b2a8c 2022-08-17)
binary: rustc
commit-hash: 9c20b2a8cc7588decb6de25ac6a7912dcef24d65
commit-date: 2022-08-17
host: x86_64-unknown-linux-gnu
release: 1.65.0-nightly
LLVM version: 15.0.0
```

**Comment by 5225225** [comment]
searched nightlies: from nightly-2022-01-01 to nightly-2022-08-16
regressed nightly: nightly-2022-08-05
searched commit range: https://github.com/rust-lang/rust/compare/1b57946a405d5b2a87e612335db033edb2c3427f...f6f9d5e73d5524b6281c10a5c89b7db35c330634
regressed commit: https://github.com/rust-lang/rust/commit/caee496150a551fe1b9f77b3a58f7e66d54bc824

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.6.4</summary>


Host triple: x86_64-unknown-linux-gnu
Reproduce with:
```bash
cargo bisect-rustc 2022-01-01 --preserve -- build 
```
</details>

Probably https://github.com/rust-lang/rust/pull/98835 ?

**Comment by steffahn** [comment]
Yes, that’s definitely #98835, but an interesting new case AFAICT.

**Comment by steffahn** [comment]
So for some context/explanation, the compiler used to (arguably somewhat “incorrectly”) act more like as if the closure was
```rs
        byte_list.iter_mut().find_map(|item| {
            Bigger::<'_>::other(item);
            Some(())
        });
```
instead of
```rs
        byte_list.iter_mut().find_map(|item| {
            Bigger::<'a>::other(item); // <- expansion of the `Self::other`
            Some(())
        });
```
and hence accepted `item` to be _shorter-lived_ than `'a`.

**Comment by Andrepuel** [comment]
> So for some context/explanation, the compiler used to (arguably somewhat “incorrectly”) act more like as if the closure was

In fact, I am glad that the new version of the compiler flagged that error. That helped me into fixing an unnecessary overly-specific lifetime constraint that slipped through the code review process.

I've reported this as the regression worried me.

**Comment by steffahn** [comment]
> I've reported this as the regression worried me.

Rightfully so, better be safe than sorry.

> In fact, I am glad that the new version of the compiler flagged that error.

FYI, I don’t think removing the compiler error entirely will happen, so even if this kind of regression is addressed, the usefulness of the compiler will not be reduced; as you noted it’s a useful error message, (and also the new check is there to prevent lots of *new* cases that would otherwise compiler successfully with the newly enabled full NLL-only borrow checking). AFAICT, the only thing still up to discussion is whether or not perhaps the new error should initially only be a lint. Possibly deny-by-default (so it’ll look a lot like an error, but not break downstream code depending on broken crates), and most definitely a future-incompatibility lint, so you (or your dependencies) cannot accidentally (i.e. without noticing it) write new code with such lifetime-annotation-inaccuracies that might be broken later if we *do* make it a hard error later. (Perhaps also with improved diagnostics to help fixing the issue.)

**Comment by aliemjay** [comment]
error output under #100976 
```
  --> /tmp/tto6.rs:15:9
   |
8  | impl<'a> Bigger<'a> {
   |      -- lifetime `'a` defined here
9  |     pub fn get_addr(byte_list: &'a mut Vec<u8>) -> &mut u8 {
10 |         byte_list.iter_mut().find_map(|item| {
   |         -------------------- first mutable borrow occurs here
11 |             Self::other(item); // replace with `Bigger`
   |             ----------- type annotation requires that `*byte_list` is borrowed for `'a`
...
15 |         byte_list.push(0);
   |         ^^^^^^^^^^^^^^^^^ second mutable borrow occurs here
   |
help: you can remove unnecessary lifetime annotations here
  --> /tmp/tto6.rs:11:13
   |
11 |             Self::other(item); // replace with `Bigger`
   |             ^^^^^^^^^^^
   = help: consider replacing `Self` with `Bigger`
```

**Comment by apiraino** [comment]
WG-prioritization assigning priority ([Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.23100725.20Borrow.20checking.20for.20static.20methods.20became.20more.20stri.E2.80.A6)).

@rustbot label -I-prioritize +P-high

**Comment by pnkfelix** [comment]
[Visited during P-high 2022 Q3 review.](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bsteering.20meeting.5D.202022-10-28.20compiler-team.23559.20P-high.20II/near/306666295)

We agreed that this regression was expected, and the answer here is improving the diagnostics.

Also, we note that this should be resolved when PR #100976

We're lowering the priority from P-high to P-medium.

@rustbot label: -P-high +P-medium

**Comment by pnkfelix** [comment]
Also reassigning to T-types

@rustbot label: -T-compiler +T-types
